### DISCUSSION TITLE
DISPLACEMENTS offset syntax

### DISCUSSION THREAD

**Discussion by alexmiame** [viperleed/viperleed#280] [Ideas]
I am working on our new tensor-LEED implementation and recently started adapting the DISPLACEMENTS file parser to work with it. While doing that, some questions came up about our current syntax with regard to defining parameter offsets.
After a thorough discussion with @michele-riva, we concluded that offset definitions in search blocks can be somewhat ambiguous, especially when there are multiple (or looped) searches. Take the following case:

```
== SEARCH z
  = GEO_DELTA
  Ir 1 z = -0.05 0.05 0.01
  Ir 1 z offset = 0.02

== SEARCH x
  = GEO_DELTA
  Ir 1 xy[1 0] = -0.03 0.03 0.01
  Ir 1 xy[1 0] offset = 0.02
```

In our current implementation, this would first offset the Ir atom in z, then optimize z, offset in x, and finally optimize x. However, the user might expect (and want) the atom to be offset in both z&x directions before either optimization is started.
Furthermore, if this search were to be looped, the optimized positions for the Ir atom is discarded entirely and reset to the original offset position.

Our suggestion would be as follows:

1. Deprecate offsets in search blocks and replace them with a new (optional) `= OFFSETS` block at the beginning of the DISPLACEMENTS file. Any offsets must be defined there and will be applied to the structure & ranges at the beginning of the optimization. The syntax for this block would be similar to the one used by the `= CONSTRAIN` block. This removes any ambiguities about how and when offsets are applied. An added benefit would be that this syntax offers a natural way to request single-shot tensor-LEED evaluations without running an optimization (i.e. invoking a superpos).

2. Deprecate and disallow resetting offsets between search runs (i.e. lines like `O 1 offset = clear`). These commands essentially allow a parameter to "snap back" to its original position after a search run. We can't really see any cases where this procedure would be beneficial, and it seems rather confusing instead.

Would be great to hear an opinion by @fkraushofer :)

**Comment by fkraushofer** [comment]
tl;dr: Sounds good.

> In our current implementation, this would first offset the Ir atom in z, then optimize z, offset in x, and finally optimize x. However, the user might expect (and want) the atom to be offset in both z&x directions before either optimization is started. 

I think it would be quite stupid to expect that, as the general behaviour of DISPLACEMENTS, i.e. that blocks are done in series, is pretty obvious. Concerning "want": Defining both displacements in the first block would be allowed, according to the documentation. This is equivalent to your proposed solution, but I agree that it is less obvious.

>Furthermore, if this search were to be looped, the optimized positions for the Ir atom is discarded entirely and reset to the original offset position.

This is indeed a major issue. One way to deal with this would be to ignore offsets in all but the first loop iteration, but I fully agree that it also doesn't make much sense to introduce a new displacement after some block has already been done. At least, I can't think of a sensible use case. So only allowing displacements at the very beginning seems sensible to me!

>An added benefit would be that this syntax offers a natural way to request single-shot tensor-LEED evaluations without running an optimization (i.e. invoking a superpos).

This one is not entirely clear to me. How would you then distinguish between this and the general case where an actual search block follows (in which case deltas shouldn't be calculated directly after the OFFSETS block)? Simply read to end of file, and if only offsets are there, then do the single-step case?

**Comment by michele-riva** [comment]
A few words to perhaps clarify the original post by @amimre:

```
== SEARCH z
  = GEO_DELTA
  Ir 1 z = -0.05 0.05 0.01
  Ir 1 z offset = 0.02         << this is currently not allowed: same direction as displacement

== SEARCH x
  = GEO_DELTA
  Ir 1 xy[1 0] = -0.03 0.03 0.01
  Ir 1 xy[1 0] offset = 0.02   << this is currently not allowed: same direction as displacement
```

The offsets that we currently allow to be specified are in a **different** direction than the one specified for one block. Something like this would however be currently allowed, but I feel it does confuse more than anything:

```
== SEARCH z
  = GEO_DELTA
  Ir 1 z = -0.05 0.05 0.01
  Ir 1 xy[1 0] offset = 0.02

== SEARCH x
  = GEO_DELTA
  Ir 1 xy[1 0] = -0.03 0.03 0.01
  Ir 1 z offset = 0.02
```

In the first search, The Ir atom is displaced along `x`, and a search runs along `z` (fine, nothing weird). In the second block that starts at the optimized position of the first one the search is along `x` (which was **shifted in the first block**) and the offset is along `z`, **away from the optimal position**. This seems very counterintuitive, if not dangerous.

**Comment by michele-riva** [comment]
> > An added benefit would be that this syntax offers a natural way to request single-shot tensor-LEED evaluations without running an optimization (i.e. invoking a superpos).
> 
> This one is not entirely clear to me. How would you then distinguish between this and the general case where an actual search block follows (in which case deltas shouldn't be calculated directly after the OFFSETS block)? Simply read to end of file, and if only offsets are there, then do the single-step case?

Yes

**Comment by michele-riva** [comment]
More to it: we would allow only a single `= OFFSET_DELTA` block, at the beginning of the DISPLACEMENTS file. This prevents it from being forgotten in the middle of a file.

**Comment by fkraushofer** [comment]
> The offsets that we currently allow to be specified are in a **different** direction than the one specified for one block. Something like this would however be currently allowed, but I feel it does confuse more than anything:
> 
> ```
> == SEARCH z
>   = GEO_DELTA
>   Ir 1 z = -0.05 0.05 0.01
>   Ir 1 xy[1 0] offset = 0.02
> 
> == SEARCH x
>   = GEO_DELTA
>   Ir 1 xy[1 0] = -0.03 0.03 0.01
>   Ir 1 z offset = 0.02
> ```
> 
> In the first search, The Ir atom is displaced along `x`, and a search runs along `z` (fine, nothing weird). In the second block that starts at the optimized position of the first one the search is along `x` (which was **shifted in the first block**) and the offset is along `z`, **away from the optimal position**. This seems very counterintuitive, if not dangerous.

Agree about the fact that allowing displacement in the second block doesn't really make sense (at least not in any usecase I can think of).

About offsets in the same direction as displacements: There is no real reason to not allow them, except that it's likely a mistake, since the same thing can be done by simply shifting the displacement range:

```
  Ir 1 z = -0.05 0.05 0.01
  Ir 1 z offset = 0.02
# is equivalent to
  Ir 1 z = -0.03 0.07 0.01
```
However, if the atoms are referenced not by "Ir 1" but by some wildcard expression, it may easily be that someone tries assigning offsets to one group of atoms, displacements to another, and some atoms are in both groups - there may be a reasonable usecase for this somewhere (e.g. offsets for different elements in one site), so perhaps we should allow it. But as I understand it, in your proposed implementation, it would be allowed anyway.

**Comment by michele-riva** [comment]
> About offsets in the same direction as displacements: There is no real reason to not allow them, except that it's likely a mistake, since the same thing can be done by simply shifting the displacement range:
> 
> ```
>   Ir 1 z = -0.05 0.05 0.01
>   Ir 1 z offset = 0.02
> # is equivalent to
>   Ir 1 z = -0.03 0.07 0.01
> ```
> 
> However, if the atoms are referenced not by "Ir 1" but by some wildcard expression, it may easily be that someone tries assigning offsets to one group of atoms, displacements to another, and some atoms are in both groups - there may be a reasonable usecase for this somewhere (e.g. offsets for different elements in one site), so perhaps we should allow it. But as I understand it, in your proposed implementation, it would be allowed anyway.

Agreed. There's no reason not to allow them. They would only apply to the "initial configuration" at the beginning of the very first search.

**Comment by michele-riva** [comment]
Another advantage: we would internally treat `SEARCH_OFFSETS` in `VIBROCC` in an identical manner as the `OFFSET_DELTA`, with the difference that the latter have a more user-friendly syntax.

**Comment by alexmiame** [comment]
Alright, seems like we are on the same page then.
I will work on implementing this new syntax for our work-in-progress new tensor-LEED implementation and we can port it to the main repository from there. Expect it to be a while though.

**Comment by alexmiame** [comment]
Work on this is in progress; I'm converting this issue into a discussion as it fits better there.
